core/trino-main/src/main/java/io/trino/likematcher/DFA.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.likematcher;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+
+record DFA(State start, State failed, List<State> states, Map<Integer, List<Transition>> transitions)
```

Comment:
Still deciding whether this should be a Record or a Class. The issue is that for a Record, the `states` and `transitions` will reference the provided instances w/o an intervening defensive copy or immutability guarantees.

Reply:
I'm concerned that it requires that extra code in construction, which seems a bit unnatural.

Reply:
Making it a Class doesn't let you get away from that.
Alternatively you could expect user to provide `ImmutableList` and `ImmutableMap` explicitly.
I wouldn't mind personally, but it would be a deviation from what we have practiced so far.